File: eval-golden.mjs
#!/usr/bin/env node
/**
 * eval-golden.mjs — golden-set eval harness for cheap-model routing (#1354)
 *
 * SCAFFOLDING. The *mechanism* in this file is design-invariant: load labeled
 * golden cases, obtain each candidate model's `---SCORE_SUMMARY---` block
 * (replayed from a recorded fixture for $0 deterministic CI, or live via
 * openai-eval.mjs), compare it to the reference label, and exit 0/1 on an
 * aggregate threshold. That part should hold regardless of the open design
 * questions on #1354.
 *
 * What is deliberately left as TODO(#1354) pending maintainer steer — these are
 * the four calls raised in the issue thread, surfaced as named constants /
 * placeholder data so they are trivial to tune once decided:
 *   - reference labels   → the `label` blocks in evals/golden/*.json
 *   - SCORE agreement     → SCORE_TOLERANCE (exact vs band)
 *   - CI gate threshold   → MIN_ARCHETYPE_AGREEMENT
 *   - per-model $/run      → COST_PER_RUN_USD
 *
 * Usage:
 *   node eval-golden.mjs --replay --model cheap-stub     # offline, deterministic ($0)
 *   node eval-golden.mjs --live   --model gpt-4o-mini    # calls openai-eval.mjs (needs key + cv.md)
 *   npm run eval:golden -- --replay --model cheap-stub
 */

import { readFileSync, readdirSync, existsSync, writeFileSync, mkdtempSync, rmSync } from 'fs';
import { join, dirname } from 'path';
import { tmpdir } from 'os';
import { fileURLToPath } from 'url';
import { spawnSync } from 'child_process';

const ROOT = dirname(fileURLToPath(import.meta.url));
const GOLDEN_DIR = join(ROOT, 'evals', 'golden');

// ---------------------------------------------------------------------------
// TODO(#1354): the four open design calls, surfaced as tunable constants.
// ---------------------------------------------------------------------------

/** Max |candidate.score - label.score| still counted as agreement.
 *  TODO(#1354): maintainer to confirm exact-match (0) vs a tolerance band. */
const SCORE_TOLERANCE = 0.5;

/** Fraction of cases whose ARCHETYPE must match the label for the gate to pass.
 *  ARCHETYPE exact-match is the clean 0/1 signal hinted at in the issue.
 *  TODO(#1354): maintainer to confirm the CI threshold. */
const MIN_ARCHETYPE_AGREEMENT = 0.8;

/** Rough $/run per model id, for the cost column. Empty until rates are agreed.
 *  TODO(#1354): populate from the providers we actually want to route to. */
const COST_PER_RUN_USD = {};

// ---------------------------------------------------------------------------
// CLI args
// ---------------------------------------------------------------------------
const args = process.argv.slice(2);
if (args.includes('--help') || args.includes('-h')) {
  console.log(`eval-golden.mjs — golden-set eval harness (#1354, scaffolding)

  --replay         Replay recorded fixtures (default; offline, $0, deterministic)
  --live           Call the model live via openai-eval.mjs (needs key + cv.md)
  --model <id>     Candidate model id to evaluate (default: cheap-stub)
  --golden <dir>   Golden-set directory (default: evals/golden)
  --fixtures <dir> Replay fixtures directory (default: sibling of --golden)
  --help           Show this help
`);
  process.exit(0);
}

const mode  = args.includes('--live') ? 'live' : 'replay';
const model = argValue('--model') || 'cheap-stub';
const goldenDir = argValue('--golden') || GOLDEN_DIR;
// Keep fixtures next to the golden set so a custom --golden dir resolves its
// own fixtures (the default lands on evals/fixtures); override with --fixtures.
const fixtureDir = argValue('--fixtures') || join(dirname(goldenDir), 'fixtures');

/**
 * Read the value following a `--flag` token in argv.
 *
 * @param {string} flag - The flag whose following value to return.
 * @returns {string|undefined} The value, or undefined if the flag is absent/last.
 */
function argValue(flag) {
  const i = args.indexOf(flag);
  return i >= 0 && i + 1 < args.length ? args[i + 1] : undefined;
}

// ---------------------------------------------------------------------------
// Shared SCORE_SUMMARY parser — same contract every *-eval.mjs already emits.
// ---------------------------------------------------------------------------

/**
 * Parse the machine-readable summary block produced by the eval scripts.
 *
 * @param {string} text - Raw model output containing a SCORE_SUMMARY block.
 * @returns {{score: number, archetype: string}} Parsed score/archetype; score
 *   is NaN and archetype is "unknown" when the block is missing or malformed.
 */
function parseSummary(text) {
  const block = text.match(/---SCORE_SUMMARY---\s*([\s\S]*?)---END_SUMMARY---/);
  const field = (key) => {
    const m = block && block[1].match(new RegExp(`${key}:\\s*(.+)`));
    return m ? m[1].trim() : '';
  };
  return {
    score:     parseFloat(field('SCORE')),
    archetype: (field('ARCHETYPE') || 'unknown').toLowerCase(),
  };
}

// ---------------------------------------------------------------------------
// Obtain one candidate completion (replay fixture or live openai-eval call).
// ---------------------------------------------------------------------------

/**
 * Return the candidate model's raw evaluation text for one golden case.
 *
 * In replay mode this reads a recorded fixture so the gate is offline and
 * deterministic; in live mode it shells out to openai-eval.mjs, reusing the
 * real prompt-assembly path rather than duplicating it here.
 *
 * @param {{id: string, jd: string}} testCase - The golden case being run.
 * @returns {string} Raw model output (expected to contain a SCORE_SUMMARY block).
 */
function getCompletion(testCase) {
  if (mode === 'replay') {
    const fixture = join(fixtureDir, `${testCase.id}__${model}.txt`);
    if (!existsSync(fixture)) {
      throw new Error(`missing replay fixture: ${fixture} — record it or run --live`);
    }
    return readFileSync(fixture, 'utf8');
  }

  // live: write the JD to a temp file and run the existing evaluator.
  const dir = mkdtempSync(join(tmpdir(), 'eval-golden-'));
  try {
    const jdFile = join(dir, 'jd.txt');
    writeFileSync(jdFile, testCase.jd);
    const res = spawnSync(process.execPath,
      [join(ROOT, 'openai-eval.mjs'), '--file', jdFile, '--model', model, '--no-save'],
      { encoding: 'utf8', env: process.env, timeout: 360000 });
    if (res.status !== 0) {
      throw new Error(`openai-eval.mjs exited ${res.status}: ${(res.stderr || '').slice(0, 200)}`);
    }
    return res.stdout || '';
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

/**
 * Median of a numeric array (0 for an empty array).
 *
 * @param {number[]} xs - Values to summarize.
 * @returns {number} The median value.
 */
function median(xs) {
  if (xs.length === 0) return 0;
  const s = [...xs].sort((a, b) => a - b);
  const mid = Math.floor(s.length / 2);
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
}

// ---------------------------------------------------------------------------
// Run
// ---------------------------------------------------------------------------
if (!existsSync(goldenDir)) {
  console.error(`❌  golden-set directory not found: ${goldenDir}`);
  process.exit(1);
}

let cases;
try {
  cases = readdirSync(goldenDir)
    .filter((f) => f.endsWith('.json'))
    .map((f) => {
      const parsed = JSON.parse(readFileSync(join(goldenDir, f), 'utf8'));
      if (typeof parsed?.id !== 'string' ||
          typeof parsed?.jd !== 'string' ||
          typeof parsed?.label?.archetype !== 'string' ||
          typeof parsed?.label?.score !== 'number') {
        throw new Error(`invalid golden case ${f}: need string id/jd and label.{archetype:string, score:number}`);
      }
      return parsed;
    });
} catch (err) {
  console.error(`❌  ${err.message || err}`);
  process.exit(1);
}

if (cases.length === 0) {
  console.error(`❌  no golden cases (*.json) in ${goldenDir}`);
  process.exit(1);
}

console.log(`\ngolden-set eval — model "${model}" (${mode}), ${cases.length} case(s)\n`);

let archetypeHits = 0;
const deltas = [];
const latencies = [];

for (const tc of cases) {
  const t0 = Date.now();
  let parsed;
  try {
    parsed = parseSummary(getCompletion(tc));
  } catch (err) {
    console.log(`  ❌ ${tc.id}: ${err.message}`);
    deltas.push(NaN);
    continue;
  }
  const latencyMs = Date.now() - t0;
  latencies.push(latencyMs);

  const archetypeMatch = parsed.archetype === String(tc.label.archetype).toLowerCase();
  const delta = Math.abs(parsed.score - tc.label.score);
  const scoreOk = Number.isFinite(delta) && delta <= SCORE_TOLERANCE;
  if (archetypeMatch) archetypeHits++;
  deltas.push(delta);

  const ok = archetypeMatch && scoreOk;
  console.log(
    `  ${ok ? '✅' : '❌'} ${tc.id}: ` +
    `archetype ${parsed.archetype} vs ${String(tc.label.archetype).toLowerCase()} ` +
    `(${archetypeMatch ? 'match' : 'MISS'}); ` +
    `score ${parsed.score} vs ${tc.label.score} (Δ${Number.isFinite(delta) ? delta.toFixed(2) : 'n/a'}); ` +
    `${mode === 'live' ? `${latencyMs}ms` : 'replay'}`,
  );
}

const agreement = archetypeHits / cases.length;
const finiteDeltas = deltas.filter(Number.isFinite);
const meanDelta = finiteDeltas.length ? finiteDeltas.reduce((a, b) => a + b, 0) / finiteDeltas.length : NaN;
const cost = COST_PER_RUN_USD[model];

console.log('\n  ── summary ──');
console.log(`  archetype agreement : ${(agreement * 100).toFixed(0)}%  (gate ≥ ${(MIN_ARCHETYPE_AGREEMENT * 100).toFixed(0)}%)`);
console.log(`  mean |Δscore|       : ${Number.isFinite(meanDelta) ? meanDelta.toFixed(2) : 'n/a'}  (tolerance ±${SCORE_TOLERANCE})`);
if (mode === 'live') console.log(`  median latency      : ${median(latencies)}ms`);
console.log(`  est. $/run          : ${cost != null ? `$${cost}` : 'n/a — TODO(#1354)'}`);

const passed = agreement >= MIN_ARCHETYPE_AGREEMENT;
console.log(`\n  ${passed ? '✅ PASS' : '❌ FAIL'} — archetype agreement ${passed ? 'meets' : 'below'} gate\n`);
process.exit(passed ? 0 : 1);

File: scripts/python/evaluation/eval_golden.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import math
import re
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Any, Callable

from scripts.python import PROJECT_ROOT


GOLDEN_DIR = PROJECT_ROOT / "evals" / "golden"
SCORE_TOLERANCE = 0.5
MIN_ARCHETYPE_AGREEMENT = 0.8
COST_PER_RUN_USD: dict[str, float] = {}


def fixture_model_id(model: str) -> str:
    return re.sub(r"[^A-Za-z0-9._-]+", "-", model)


def parse_summary(text: str) -> dict[str, Any]:
    block = re.search(r"---SCORE_SUMMARY---\s*([\s\S]*?)---END_SUMMARY---", text)

    def field(key: str) -> str:
        if not block:
            return ""
        match = re.search(rf"{re.escape(key)}:\s*(.+)", block.group(1))
        return match.group(1).strip() if match else ""

    try:
        score = float(field("SCORE"))
    except ValueError:
        score = math.nan
    return {"score": score, "archetype": (field("ARCHETYPE") or "unknown").lower()}


def median(values: list[float]) -> float:
    if not values:
        return 0
    sorted_values = sorted(values)
    mid = len(sorted_values) // 2
    return sorted_values[mid] if len(sorted_values) % 2 else (sorted_values[mid - 1] + sorted_values[mid]) / 2


def load_cases(golden_dir: str | Path) -> list[dict[str, Any]]:
    directory = Path(golden_dir)
    if not directory.exists():
        raise FileNotFoundError(f"golden-set directory not found: {directory}")
    cases: list[dict[str, Any]] = []
    for path in sorted(directory.glob("*.json")):
        parsed = json.loads(path.read_text(encoding="utf-8"))
        label = parsed.get("label") if isinstance(parsed.get("label"), dict) else {}
        if not isinstance(parsed.get("id"), str) or not isinstance(parsed.get("jd"), str) or not isinstance(label.get("archetype"), str) or not isinstance(label.get("score"), (int, float)):
            raise ValueError(f"invalid golden case {path.name}: need string id/jd and label.{{archetype:string, score:number}}")
        cases.append(parsed)
    if not cases:
        raise ValueError(f"no golden cases (*.json) in {directory}")
    return cases


def replay_completion(test_case: dict[str, Any], *, fixture_dir: str | Path, model: str) -> str:
    fixture = Path(fixture_dir) / f"{test_case['id']}__{fixture_model_id(model)}.txt"
    if not fixture.exists():
        raise FileNotFoundError(f"missing replay fixture: {fixture} — record it or run --live")
    return fixture.read_text(encoding="utf-8")


def live_completion(test_case: dict[str, Any], *, model: str, root: str | Path = PROJECT_ROOT) -> str:
    with tempfile.TemporaryDirectory(prefix="eval-golden-") as directory:
        jd_file = Path(directory) / "jd.txt"
        jd_file.write_text(test_case["jd"], encoding="utf-8")
        result = subprocess.run(
            [
                "python",
                "-m",
                "scripts.python.evaluation.openai_eval",
                "--file",
                str(jd_file),
                "--model",
                model,
                "--no-save",
            ],
            cwd=Path(root),
            capture_output=True,
            text=True,
            timeout=360,
        )
        if result.returncode != 0:
            raise RuntimeError(f"openai_eval exited {result.returncode}: {(result.stderr or result.stdout)[:200]}")
        return result.stdout


def evaluate_cases(
    cases: list[dict[str, Any]],
    *,
    model: str,
    mode: str = "replay",
    fixture_dir: str | Path | None = None,
    completion_provider: Callable[[dict[str, Any]], str] | None = None,
) -> dict[str, Any]:
    rows: list[dict[str, Any]] = []
    archetype_hits = 0
    deltas: list[float] = []
    latencies: list[float] = []
    for case in cases:
        start = time.perf_counter()
        try:
            if completion_provider:
                raw = completion_provider(case)
            elif mode == "replay":
                if fixture_dir is None:
                    raise ValueError("fixture_dir is required in replay mode")
                raw = replay_completion(case, fixture_dir=fixture_dir, model=model)
            else:
                raw = live_completion(case, model=model)
            parsed = parse_summary(raw)
            latency_ms = round((time.perf_counter() - start) * 1000)
            latencies.append(latency_ms)
            expected_archetype = str(case["label"]["archetype"]).lower()
            archetype_match = parsed["archetype"] == expected_archetype
            delta = abs(parsed["score"] - float(case["label"]["score"]))
            score_ok = math.isfinite(delta) and delta <= SCORE_TOLERANCE
            if archetype_match:
                archetype_hits += 1
            deltas.append(delta)
            rows.append(
                {
                    "id": case["id"],
                    "ok": archetype_match and score_ok,
                    "archetype": parsed["archetype"],
                    "expectedArchetype": expected_archetype,
                    "archetypeMatch": archetype_match,
                    "score": parsed["score"],
                    "expectedScore": float(case["label"]["score"]),
                    "delta": delta,
                    "scoreOk": score_ok,
                    "latencyMs": latency_ms,
                }
            )
        except Exception as error:
            deltas.append(math.nan)
            rows.append({"id": case["id"], "ok": False, "error": str(error), "delta": math.nan})
    agreement = archetype_hits / len(cases)
    finite_deltas = [delta for delta in deltas if math.isfinite(delta)]
    mean_delta = sum(finite_deltas) / len(finite_deltas) if finite_deltas else math.nan
    passed = agreement >= MIN_ARCHETYPE_AGREEMENT
    return {
        "model": model,
        "mode": mode,
        "rows": rows,
        "summary": {
            "archetypeAgreement": agreement,
            "meanScoreDelta": mean_delta,
            "scored": len(finite_deltas),
            "total": len(cases),
            "unscored": len(cases) - len(finite_deltas),
            "medianLatencyMs": median(latencies),
            "costPerRunUsd": COST_PER_RUN_USD.get(model),
            "passed": passed,
        },
    }


def format_report(result: dict[str, Any]) -> str:
    lines = [f'\ngolden-set eval — model "{result["model"]}" ({result["mode"]}), {len(result["rows"])} case(s)', ""]
    for row in result["rows"]:
        if row.get("error"):
            lines.append(f"  FAIL {row['id']}: {row['error']}")
            continue
        delta = row["delta"]
        lines.append(
            f"  {'OK' if row['ok'] else 'FAIL'} {row['id']}: "
            f"archetype {row['archetype']} vs {row['expectedArchetype']} "
            f"({'match' if row['archetypeMatch'] else 'MISS'}); "
            f"score {row['score']} vs {row['expectedScore']} "
            f"(delta {delta:.2f} if finite)"
        )
    summary = result["summary"]
    mean_delta = summary["meanScoreDelta"]
    lines.extend(
        [
            "",
            "  -- summary --",
            f"  archetype agreement : {summary['archetypeAgreement'] * 100:.0f}%  (gate >= {MIN_ARCHETYPE_AGREEMENT * 100:.0f}%)",
            f"  mean |delta score|  : {mean_delta:.2f}" if math.isfinite(mean_delta) else "  mean |delta score|  : n/a",
            f"  scored              : {summary['scored']}/{summary['total']} ({summary['unscored']} unscored)",
            f"  {'PASS' if summary['passed'] else 'FAIL'}",
        ]
    )
    return "\n".join(lines) + "\n"


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Golden-set eval harness for cheap-model routing.")
    mode = parser.add_mutually_exclusive_group()
    mode.add_argument("--replay", action="store_true")
    mode.add_argument("--live", action="store_true")
    parser.add_argument("--model", default="cheap-stub")
    parser.add_argument("--golden", default=str(GOLDEN_DIR))
    parser.add_argument("--fixtures")
    parser.add_argument("--json", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    mode = "live" if args.live else "replay"
    fixture_dir = args.fixtures or str(Path(args.golden).parent / "fixtures")
    try:
        cases = load_cases(args.golden)
        result = evaluate_cases(cases, model=args.model, mode=mode, fixture_dir=fixture_dir)
    except Exception as error:
        print(json.dumps({"error": str(error)}, indent=2) if args.json else f"ERROR: {error}")
        return 1
    print(json.dumps(result, indent=2) if args.json else format_report(result))
    return 0 if result["summary"]["passed"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
